Mage.Sets/src/mage/cards/f/FireGiantsFury.java

```diff
@@ -0,0 +1,242 @@
+
+    private static final FilterControlledCreaturePermanent filter = new FilterControlledCreaturePermanent(SubType.GIANT);
+
+    public FireGiantsFury(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.SORCERY}, "{1}{R}");
+
+        // Target Giant you control gets +2/+2 and gains trample until end of turn.
+        Effect effect = new BoostTargetEffect(2, 2, Duration.EndOfTurn);
+        effect.setText("Target Giant you control gets +2/+2");
+        this.getSpellAbility().addEffect(effect);
+
+        effect = new GainAbilityTargetEffect(TrampleAbility.getInstance(), Duration.EndOfTurn);
+        effect.setText("and gains trample until end of turn");
+        this.getSpellAbility().addEffect(effect);
+
+        // Whenever it deals combat damage to a player this turn, exile that many cards from the top of your library. Until the end of your next turn, you may play those cards.
+        effect = new FireGiantsFuryEffect();
+        this.getSpellAbility().addEffect(effect);
+
+        this.getSpellAbility().addTarget(new TargetPermanent(filter));
+    }
+
+    private FireGiantsFury(final FireGiantsFury card) {
+        super(card);
+    }
+
+    @Override
+    public FireGiantsFury copy() {
+        return new FireGiantsFury(this);
+    }
+}
+
+class FireGiantsFuryEffect extends OneShotEffect {
+
+    public FireGiantsFuryEffect() {
+        super(Outcome.PlayForFree);
+        this.staticText = "Whenever it deals combat damage to a player this turn, exile that many cards from the top of your library. Until the end of your next turn, you may play those cards";
+    }
+
+    private FireGiantsFuryEffect(final FireGiantsFuryEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public FireGiantsFuryEffect copy() {
+        return new FireGiantsFuryEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        DelayedTriggeredAbility delayedAbility = new FireGiantsFuryDelayedTriggeredAbility(source.getFirstTarget());
+        delayedAbility.addWatcher(new FireGiantsFuryWatcher());
+        game.addDelayedTriggeredAbility(delayedAbility, source);
+        return true;
+    }
+}
+
+class FireGiantsFuryDelayedTriggeredAbility extends DelayedTriggeredAbility {
+
+    private final UUID target;
+
+    public FireGiantsFuryDelayedTriggeredAbility(UUID target) {
+        super(new FireGiantsFuryDelayedEffect(target), Duration.EndOfTurn, false, false);
+        this.target = target;
+    }
+
+    private FireGiantsFuryDelayedTriggeredAbility (FireGiantsFuryDelayedTriggeredAbility ability) {
+        super(ability);
+        this.target = ability.target;
+    }
+
+    @Override
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.DAMAGED_PLAYER;
+    }
+
+    @Override
+    public boolean checkTrigger(GameEvent event, Game game) {
+        return event.getSourceId().equals(target);
```

Comment:
I've got it re-implemented as a delayed triggered ability now.  Only problem I found in testing is that it now applies to both combat and non-combat damage (ex. Sorcerer's Wand.)

I tried `return (event.getSourceId().equals(target) && event.getFlag());` here (saw a comment in GameEvent.Java that said flag should be true for combat damage) but doing this just stopped the ability from triggering entirely.

Reply:
You have to cast it to `DamagedEvent` and use `isCombatDamage()`

Reply:
Also you don't need the watcher, you can pass the damage amount to the effect in the `checkTrigger` method